### ISSUE TITLE
Panic when there are stopped containers in a Fargate task

### ISSUE THREAD

**Issue by zlangbert** [DataDog/datadog-process-agent#165]
I believe the process agent is suffering from the same issue described here: DataDog/integrations-core#1955.

Short version is that the container stats endpoint returns a key for stopped containers but a `null` stats object. The process agent appears to assume that there is data there.

You can see an example of metadata and stats output triggering this [here](https://gist.github.com/zlangbert/41540dd857dbbb35ba6831d32f95968b).

Agent log:
```
[PROCESS] starting process-agent
[PROCESS] Error: unable to determine creation time for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - parsing time "" as "2006-01-02T15:04:05Z07:00": cannot parse "" as "2006"
[PROCESS] Error: unable to get stats from ECS for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] Error: unable to get stats from ECS for container 75aba7c191d45c9121f0ae4a7835c6dc4e047e184adc54610c00e6d500bdb55f - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] 2018-08-17 20:58:51 INFO (config.go:398) - overriding API key from env DD_API_KEY value
[PROCESS] Error: unable to get stats from ECS for container 75aba7c191d45c9121f0ae4a7835c6dc4e047e184adc54610c00e6d500bdb55f - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] Error: unable to determine creation time for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - parsing time "" as "2006-01-02T15:04:05Z07:00": cannot parse "" as "2006"
[PROCESS] Error: unable to get stats from ECS for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] Error: unable to determine creation time for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - parsing time "" as "2006-01-02T15:04:05Z07:00": cannot parse "" as "2006"
[PROCESS] Error: unable to get stats from ECS for container 5862d65c8843b4ffc2ebb8e4437aa4e49f24e3e2c5b89ce22d9bd8854522eb25 - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] Error: unable to get stats from ECS for container 75aba7c191d45c9121f0ae4a7835c6dc4e047e184adc54610c00e6d500bdb55f - json: cannot unmarshal string into Go value of type ecs.ContainerStats
[PROCESS] panic: runtime error: invalid memory address or nil pointer dereference
[PROCESS] [signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x11035a6]
[PROCESS]
[PROCESS] goroutine 41 [running]:
[PROCESS] github.com/DataDog/datadog-process-agent/checks.fmtContainers(0xc4201c0960, 0x5, 0x6, 0xc4204c57a0, 0x5, 0x6, 0xbed5eba2df6405b7, 0x1576c8a0, 0x1fafbe0, 0x1, ...)
[PROCESS] /home/jenkins/workspace/process-agent-build-ddagent/go/src/github.com/DataDog/datadog-process-agent/checks/container.go:111 +0x356
[PROCESS] github.com/DataDog/datadog-process-agent/checks.(*ContainerCheck).Run(0x1faf660, 0xc420339380, 0x61278774, 0xc420339e00, 0xc42031df10, 0xc42008da40, 0xc420097980, 0xc42008da40)
[PROCESS] /home/jenkins/workspace/process-agent-build-ddagent/go/src/github.com/DataDog/datadog-process-agent/checks/container.go:63 +0x135
[PROCESS] main.(*Collector).runCheck(0xc42033ac80, 0x1541f80, 0x1faf660)
[PROCESS] /home/jenkins/workspace/process-agent-build-ddagent/go/src/github.com/DataDog/datadog-process-agent/agent/collector.go:78 +0xb4
[PROCESS] main.(*Collector).run.func2(0xc42033ac80, 0xc420096f00, 0x1541f80, 0x1faf660)
[PROCESS] /home/jenkins/workspace/process-agent-build-ddagent/go/src/github.com/DataDog/datadog-process-agent/agent/collector.go:144 +0x310
[PROCESS] created by main.(*Collector).run
[PROCESS] /home/jenkins/workspace/process-agent-build-ddagent/go/src/github.com/DataDog/datadog-process-agent/agent/collector.go:132 +0x3f1
[PROCESS] trace-agent exited with code 2, signal 0, restarting in 2 seconds
```

**Comment by CharlyF** [comment]
Hey @hatappi,

Thank you very much for raising this - I have been digging and I think the solution is more complex than just dropping the other types than `NORMAL` - Indeed the ECS internals in this case need to be monitored.
 @zlangbert's output confirms this, for the `fg-proxy:tinyproxy` container.
(container: `494af760157`, although of type `CNI_PAUSE` is exposing statistics)

The underlying issue is that containers with empty stats should not be considered - And I think part of the solution is to look at the containers' status to avoid collecting metrics from stopped containers as alluded to in https://github.com/DataDog/integrations-core/issues/1955.

The issue has been raised internally as well, so we can solve it asap.

Thank you very much for your patience.
Best Regards,

**Comment by truthbk** [comment]
So from:
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x11035a6]
```
It doesn't seem we're dereferencing a nil pointer, notice the addr (`0x28`). It may potentially still be a nil pointer dereference as we don't have all the arguments to `fmtContainers()`:
```
[PROCESS] github.com/DataDog/datadog-process-agent/checks.fmtContainers(0xc4201c0960, 0x5, 0x6, 0xc4204c57a0, 0x5, 0x6, 0xbed5eba2df6405b7, 0x1576c8a0, 0x1fafbe0, 0x1, ...)
``` 
But none of those are `nil` either, so it's definitely not a given that we may be calling anything on a `nil` receiver .

**Comment by truthbk** [comment]
Hopefully this will address it: https://github.com/DataDog/datadog-process-agent/pull/176

**Comment by hatappi** [comment]
@CharlyF @truthbk 
Thank you for comment and fix code.
I used [datadog/agent:6.5.0-rc.3](https://hub.docker.com/r/datadog/agent/tags/) to check if an error occurred in the agent.
I confirmed that metrics can be taken.

However, [ecs_fargate.py](https://github.com/DataDog/integrations-core/blob/master/ecs_fargate/datadog_checks/ecs_fargate/ecs_fargate.py)
 seems to cause an error yet.

**Comment by zlangbert** [comment]
Thanks all. The process agent successfully starts with agent `6.5.0-rc.3`, and I can see the everything in the container explorer. 

I do wonder if it's worth capturing / showing the internal pause and volume source containers. I know the Kubernetes integration ignores it's pause containers.

@hatappi I still see the same error in the fargate check as well. I will update my issue in [integrations-core](https://github.com/DataDog/integrations-core/issues/1955).

**Comment by ofek** [comment]
@zlangbert This should be fixed in 6.5.x final via https://github.com/DataDog/integrations-core/pull/2206

Can you confirm?

**Comment by zlangbert** [comment]
@ofek All good on 6.5.1. Thanks!
